lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,643 @@
+	if exist, err := DecodeFromBytes(globalParamsEntry, rr); !exist || err != nil {
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalParamsEntryWithTxn: problem decoding GlobalParamsEntry: ")
+	}
+	return globalParamsEntry, nil
+}
+
+func DBPutSnapshotGlobalParamsEntryWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	globalParamsEntry *GlobalParamsEntry,
+	snapshotAtEpochNumber uint64,
+	blockHeight uint64,
+) error {
+	if globalParamsEntry == nil {
+		// This should never happen but is a sanity check.
+		glog.Errorf("DBPutSnapshotGlobalParamsEntryWithTxn: called with nil GlobalParamsEntry, this should never happen")
+		return nil
+	}
+	key := DBKeyForSnapshotGlobalParamsEntry(snapshotAtEpochNumber)
+	return DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, globalParamsEntry))
+}
+
+//
+// SnapshotValidatorEntry
+//
+
+type SnapshotValidatorMapKey struct {
+	SnapshotAtEpochNumber uint64
+	ValidatorPKID         PKID
+}
+
+func (bav *UtxoView) SnapshotCurrentValidators(snapshotAtEpochNumber uint64) error {
+	// First, snapshot any !isDeleted ValidatorEntries in the UtxoView.
+	var utxoViewValidatorPKIDs []*PKID
+	for _, validatorEntry := range bav.ValidatorPKIDToValidatorEntry {
+		if !validatorEntry.isDeleted {
+			// We only want to snapshot !isDeleted ValidatorEntries.
+			bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+		}
+		// We don't want to retrieve any ValidatorEntries from the db that are present in the UtxoView.
+		utxoViewValidatorPKIDs = append(utxoViewValidatorPKIDs, validatorEntry.ValidatorPKID)
+	}
+	// Second, snapshot the ValidatorEntries in the db (skipping any in the UtxoView).
+	dbValidatorEntries, err := DBEnumerateAllCurrentValidators(bav.Handle, utxoViewValidatorPKIDs)
+	if err != nil {
+		return errors.Wrapf(err, "SnapshotValidators: problem retrieving ValidatorEntries: ")
+	}
+	for _, validatorEntry := range dbValidatorEntries {
+		bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+	}
+	return nil
+}
+
+func (bav *UtxoView) GetSnapshotValidatorByPKID(pkid *PKID, snapshotAtEpochNumber uint64) (*ValidatorEntry, error) {
+	// Check the UtxoView first.
+	mapKey := SnapshotValidatorMapKey{SnapshotAtEpochNumber: snapshotAtEpochNumber, ValidatorPKID: *pkid}
+	if validatorEntry, exists := bav.SnapshotValidatorEntries[mapKey]; exists {
+		return validatorEntry, nil
+	}
+	// If we don't have it in the UtxoView, check the db.
+	validatorEntry, err := DBGetSnapshotValidatorByPKID(bav.Handle, bav.Snapshot, pkid, snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"GetSnapshotValidatorByPKID: problem retrieving ValidatorEntry from db: ",
+		)
+	}
+	if validatorEntry != nil {
+		// Cache the result in the UtxoView.
+		bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+	}
+	return validatorEntry, nil
+}
+
+func (bav *UtxoView) GetSnapshotTopActiveValidatorsByStake(limit uint64, snapshotAtEpochNumber uint64) ([]*ValidatorEntry, error) {
```

Comment:
There's some code duplication between e.g. GetTopActiveValidatorsByStake and GetSnapshotTopActiveValidatorsByStake. Not enough to IMO refactor into one function, but just calling it out. We could have the active (non-snapshot set) default to `SnapshotAtEpochNumber = 0` and then reuse the indices + functions. So basically, if we are updating the current active validator set in consensus, we assume the `SnapshotAtEpochNumber = 0`, that's actually where we store the current validator set. And any rows in that index where `SnapshotAtEpochNumber > 1` are actually snapshot values. I think there's something nice to having them split apart (more future-proof in case their logic diverges), but I could do a pretty big refactor to combine the current + snapshot ValidatorEntries into a single index, one for by PKID and one for by stake.

Reply:
The same could be done for GlobalActiveStakeAmountNanos.